File: Bhojan-app-karthikkashyap-manager-patch-1/backend/src/admin/services/admin-employee.service.ts
import {
  Injectable,
  NotFoundException,
  ForbiddenException,
  BadRequestException,
  Logger,
} from '@nestjs/common';
import { InjectQueue } from '@nestjs/bull';
import type { Queue } from 'bull';
import { randomUUID } from 'crypto';
import { PrismaService } from '../../prisma/prisma.service.js';
import { RedisService } from '../../redis/redis.service.js';
import { AuditService } from '../../audit/services/audit.service.js';
import type { AuditLogQueryDto } from '../dto/audit-log-query.dto.js';
import type { AuditExportQueryDto } from '../dto/audit-export-query.dto.js';
import type { BulkUploadDto } from '../dto/bulk-upload.dto.js';
import type { OffboardEmployeeDto } from '../dto/offboard-employee.dto.js';
import type { SessionsQueryDto } from '../dto/sessions-query.dto.js';
import { AuditAction, AuditModule, Role, Prisma, UploadStatus } from '@prisma/client';

const REFRESH_TTL_SECONDS = 7 * 24 * 60 * 60;

@Injectable()
export class AdminEmployeeService {
  private readonly logger = new Logger(AdminEmployeeService.name);

  constructor(
    private prisma: PrismaService,
    private redis: RedisService,
    private audit: AuditService,
    @InjectQueue('employee-bulk-upload') private readonly bulkUploadQueue: Queue,
    @InjectQueue('audit-export') private readonly auditExportQueue: Queue,
  ) {}

  // ── EP-22: Admin Audit Log Query ───────────────────────────────────────────

  async getAuditLogs(
    callerRole: Role,
    callerTenantId: string | null,
    query: AuditLogQueryDto,
  ) {
    const page = query.page ?? 1;
    const limit = query.limit ?? 50;
    const skip = (page - 1) * limit;

    const where: Prisma.AuditLogWhereInput = {};

    // OPS_ADMIN can only view their own tenant
    if (callerRole === Role.OPS_ADMIN) {
      where.tenant_id = callerTenantId ?? undefined;
    } else if (query.tenant_id) {
      where.tenant_id = query.tenant_id;
    }

    if (query.user_id) where.user_id = query.user_id;
    if (query.action) where.action = query.action;
    if (query.module) where.module = query.module;

    if (query.from_date || query.to_date) {
      where.created_at = {};
      if (query.from_date) where.created_at.gte = new Date(query.from_date);
      if (query.to_date) where.created_at.lte = new Date(query.to_date);
    }

    const [total, logs] = await Promise.all([
      this.prisma.auditLog.count({ where }),
      this.prisma.auditLog.findMany({
        where,
        orderBy: { created_at: 'desc' },
        skip,
        take: limit,
        select: {
          id: true,
          user_id: true,
          tenant_id: true,
          action: true,
          module: true,
          target_id: true,
          ip_address: true,
          device: true,
          metadata: true,
          created_at: true,
        },
      }),
    ]);

    return {
      logs,
      pagination: {
        page,
        limit,
        total,
        total_pages: Math.ceil(total / limit),
      },
    };
  }

  // ── EP-26: Employee Offboarding ────────────────────────────────────────────

  async offboardEmployee(
    targetUserId: string,
    dto: OffboardEmployeeDto,
    callerUserId: string,
    callerRole: Role,
    callerTenantId: string | null,
    ip?: string,
  ) {
    const target = await this.prisma.user.findFirst({
      where: { id: targetUserId, role: Role.USER },
    });
    if (!target) {
      throw new NotFoundException({ code: 'USER_NOT_FOUND', message: 'Employee not found.' });
    }

    if (callerRole === Role.OPS_ADMIN && target.tenant_id !== callerTenantId) {
      throw new ForbiddenException({ code: 'TENANT_MISMATCH', message: 'You can only offboard employees from your own tenant.' });
    }

    // Deactivate user account
    await this.prisma.user.update({
      where: { id: targetUserId },
      data: { is_active: false, deactivated_at: new Date() },
    });

    // Deactivate roster entry
    if (target.tenant_id && target.employee_id) {
      await this.prisma.employeeRoster.updateMany({
        where: { tenant_id: target.tenant_id, employee_id: target.employee_id },
        data: { is_active: false },
      });
    }

    // Revoke all active sessions
    const activeSessions = await this.prisma.session.findMany({
      where: { user_id: targetUserId, is_revoked: false },
      select: { id: true, refresh_token: true, expires_at: true },
    });

    if (activeSessions.length > 0) {
      await this.prisma.session.updateMany({
        where: { user_id: targetUserId, is_revoked: false },
        data: { is_revoked: true },
      });

      for (const session of activeSessions) {
        const ttl = Math.max(0, Math.floor((session.expires_at.getTime() - Date.now()) / 1000));
        if (ttl > 0) {
          await this.redis.set(`blacklist:${session.refresh_token}`, '1', ttl);
        }
      }
    }

    const metadata = { performed_by: callerUserId, reason: dto.reason, sessions_revoked: activeSessions.length };

    await Promise.all([
      this.audit.log({
        userId: callerUserId, tenantId: callerTenantId ?? undefined,
        action: AuditAction.OFFBOARDING_DEACTIVATION,
        module: AuditModule.EMPLOYEE,
        targetId: targetUserId, ipAddress: ip, metadata,
      }),
      this.audit.log({
        userId: callerUserId, tenantId: target.tenant_id ?? undefined,
        action: AuditAction.EMPLOYEE_OFFBOARDED,
        module: AuditModule.EMPLOYEE,
        targetId: targetUserId, ipAddress: ip,
        metadata: { employee_id: target.employee_id, reason: dto.reason },
      }),
    ]);

    return { success: true, message: 'Employee offboarded successfully.' };
  }

  // ── EP-33: List Sessions ───────────────────────────────────────────────────

  async getSessions(query: SessionsQueryDto) {
    const where: Prisma.SessionWhereInput = {};
    if (query.user_id) where.user_id = query.user_id;
    if (query.is_active !== undefined) {
      where.is_revoked = !query.is_active;
    } else {
      where.is_revoked = false; // default: active only
    }

    const sessions = await this.prisma.session.findMany({
      where,
      orderBy: { created_at: 'desc' },
      select: {
        id: true, user_id: true, device: true, ip_address: true,
        is_revoked: true, created_at: true, expires_at: true,
        user: { select: { email: true } },
      },
    });

    return {
      sessions: sessions.map(s => ({
        id: s.id,
        user_id: s.user_id,
        user_email: s.user.email,
        device: s.device,
        ip_address: s.ip_address,
        is_revoked: s.is_revoked,
        created_at: s.created_at,
        expires_at: s.expires_at,
      })),
    };
  }

  // ── EP-34: Revoke Session ──────────────────────────────────────────────────

  async revokeSession(sessionId: string, adminId: string) {
    const session = await this.prisma.session.findUnique({ where: { id: sessionId } });
    if (!session) {
      throw new NotFoundException({ code: 'SESSION_NOT_FOUND', message: 'Session not found.' });
    }

    await this.prisma.session.update({ where: { id: sessionId }, data: { is_revoked: true } });

    const ttl = Math.max(0, Math.floor((session.expires_at.getTime() - Date.now()) / 1000));
    if (ttl > 0) {
      await this.redis.set(`blacklist:${session.refresh_token}`, '1', ttl);
    }

    await this.audit.log({
      userId: adminId, action: AuditAction.SESSION_REVOKED,
      module: AuditModule.SESSION, targetId: session.user_id,
      metadata: { session_id: sessionId },
    });

    return { success: true, message: 'Session revoked.' };
  }

  // ── EP-23: Audit logs export (async) ────────────────────────────────────────
  async exportAuditLogs(callerRole: Role, callerTenantId: string | null, query: AuditExportQueryDto) {
    if (query.job_id) {
      const status = await this.redis.get(`audit_export:${query.job_id}`);
      if (!status) throw new NotFoundException('Export job not found or expired.');
      return JSON.parse(status);
    }

    const jobId = randomUUID();
    await this.redis.set(
      `audit_export:${jobId}`,
      JSON.stringify({ job_id: jobId, status: 'PENDING', message: 'Export is being generated.' }),
      3600,
    );

    await this.auditExportQueue.add(
      'generate-export',
      {
        jobId,
        callerRole,
        callerTenantId,
        filters: query,
      },
      { attempts: 2, removeOnComplete: true, removeOnFail: 50 },
    );

    return {
      job_id: jobId,
      status: 'PENDING',
      message: 'Export request accepted. Poll this endpoint with job_id.',
    };
  }

  // ── EP-24: Bulk employee upload (async) ─────────────────────────────────────
  async uploadEmployeesCsv(
    callerRole: Role,
    callerTenantId: string | null,
    callerUserId: string,
    dto: BulkUploadDto,
    file: Express.Multer.File,
  ) {
    if (!file) throw new BadRequestException('CSV file is required.');
    if (!file.originalname.toLowerCase().endsWith('.csv')) {
      throw new BadRequestException('Only CSV files are allowed.');
    }

    const tenantId =
      callerRole === Role.SUPER_ADMIN ? dto.tenant_id ?? null : callerTenantId;
    if (!tenantId) throw new BadRequestException('tenant_id is required for SUPER_ADMIN.');

    const lines = file.buffer.toString('utf8').split(/\r?\n/).filter((line) => line.trim().length > 0);
    const totalRecords = Math.max(0, lines.length - 1);
    if (totalRecords === 0) throw new BadRequestException('CSV has no data rows.');

    const log = await this.prisma.bulkUploadLog.create({
      data: {
        tenant_id: tenantId,
        uploaded_by: callerUserId,
        file_name: file.originalname,
        total_records: totalRecords,
        status: UploadStatus.PENDING,
      },
      select: { id: true },
    });

    await this.bulkUploadQueue.add(
      'process-upload',
      {
        uploadId: log.id,
        tenantId,
        uploadedBy: callerUserId,
        csvText: file.buffer.toString('utf8'),
      },
      { attempts: 2, removeOnComplete: true, removeOnFail: 50 },
    );

    return {
      upload_id: log.id,
      status: 'PENDING',
      total_records: totalRecords,
      message: 'Upload accepted. Processing in background.',
    };
  }

  // ── EP-25: Bulk upload status poll ──────────────────────────────────────────
  async getBulkUploadStatus(id: string, callerRole: Role, callerTenantId: string | null) {
    const log = await this.prisma.bulkUploadLog.findUnique({
      where: { id },
      select: {
        id: true,
        tenant_id: true,
        status: true,
        total_records: true,
        success_count: true,
        failed_count: true,
        error_details: true,
        created_at: true,
        updated_at: true,
      },
    });
    if (!log) throw new NotFoundException('Upload job not found.');
    if (callerRole === Role.OPS_ADMIN && log.tenant_id !== callerTenantId) {
      throw new ForbiddenException('Cannot access uploads from another tenant.');
    }

    return {
      upload_id: log.id,
      status: log.status,
      total_records: log.total_records,
      success_count: log.success_count,
      failed_count: log.failed_count,
      error_details: log.error_details,
      created_at: log.created_at,
      updated_at: log.updated_at,
    };
  }
}

File: Bhojan-app-karthikkashyap-manager-patch-1/lib/admin_app/employees/services/admin_employee_service.dart
import 'package:dio/dio.dart';
import '../../../../core/api/api_client.dart';
import '../../../../core/api/api_endpoints.dart';

class AdminEmployee {
  final String id;
  final String fullName;
  final String employeeId;
  final String email;
  final String? department;
  final String tenantId;
  final bool isActive;

  AdminEmployee({
    required this.id,
    required this.fullName,
    required this.employeeId,
    required this.email,
    this.department,
    required this.tenantId,
    required this.isActive,
  });

  factory AdminEmployee.fromJson(Map<String, dynamic> json) {
    return AdminEmployee(
      id: json['id'],
      fullName: json['full_name'] ?? '',
      employeeId: json['employee_id'] ?? '',
      email: json['email'] ?? '',
      department: json['department'],
      tenantId: json['tenant_id'] ?? '',
      isActive: json['is_active'] ?? false,
    );
  }
}

class BulkUploadStatus {
  final String uploadId;
  final String status; // PENDING, PROCESSING, COMPLETED, FAILED
  final int totalRecords;
  final int successCount;
  final int failedCount;
  final List<Map<String, dynamic>> errorDetails;

  BulkUploadStatus({
    required this.uploadId,
    required this.status,
    required this.totalRecords,
    required this.successCount,
    required this.failedCount,
    required this.errorDetails,
  });

  factory BulkUploadStatus.fromJson(Map<String, dynamic> json) {
    return BulkUploadStatus(
      uploadId: json['upload_id'] ?? '',
      status: json['status'] ?? 'PENDING',
      totalRecords: json['total_records'] ?? 0,
      successCount: json['success_count'] ?? 0,
      failedCount: json['failed_count'] ?? 0,
      errorDetails: List<Map<String, dynamic>>.from(json['error_details'] ?? []),
    );
  }
}

class AdminEmployeeService {
  final ApiClient _api;

  AdminEmployeeService(this._api);

  /// EP-24: POST /admin/employees/bulk-upload — Upload CSV
  Future<BulkUploadStatus> bulkUpload({
    required List<int> fileBytes,
    required String fileName,
    String? tenantId,
  }) async {
    try {
      final formData = FormData.fromMap({
        'file': MultipartFile.fromBytes(fileBytes, filename: fileName),
        if (tenantId != null) 'tenant_id': tenantId,
      });
      final response = await _api.post(ApiEndpoints.bulkUpload, data: formData);
      return BulkUploadStatus.fromJson(response.data['data']);
    } on DioException catch (e) {
      throw _parseException(e);
    }
  }

  /// EP-25: GET /admin/employees/bulk-upload/:id — Poll upload status
  Future<BulkUploadStatus> getBulkUploadStatus(String uploadId) async {
    try {
      final response = await _api.get('${ApiEndpoints.bulkUpload}/$uploadId');
      return BulkUploadStatus.fromJson(response.data['data']);
    } on DioException catch (e) {
      throw _parseException(e);
    }
  }

  /// List employees for offboard search
  Future<List<AdminEmployee>> listEmployees({String? search, String? tenantId}) async {
    try {
      final response = await _api.get(
        '/admin/employees',
        queryParameters: {
          if (search != null && search.isNotEmpty) 'search': search,
          if (tenantId != null) 'tenant_id': tenantId,
        },
      );
      final list = response.data['data'] as List;
      return list.map((j) => AdminEmployee.fromJson(j)).toList();
    } on DioException catch (e) {
      throw _parseException(e);
    }
  }

  /// EP-26: POST /admin/employees/:id/offboard
  Future<void> offboardEmployee(String userId, String reason) async {
    try {
      await _api.post(ApiEndpoints.employeeOffboard(userId), data: {'reason': reason});
    } on DioException catch (e) {
      throw _parseException(e);
    }
  }

  Exception _parseException(DioException e) {
    final data = e.response?.data;
    String code = 'NETWORK_ERROR';
    String message = 'Failed to connect to the server.';
    if (data != null && data is Map<String, dynamic> && data['error'] != null) {
      code = data['error']['code'] ?? code;
      message = data['error']['message'] ?? message;
    }
    return AdminEmployeeException(code: code, message: message);
  }
}

class AdminEmployeeException implements Exception {
  final String code;
  final String message;
  AdminEmployeeException({required this.code, required this.message});

  String get userMessage {
    switch (code) {
      case 'TENANT_MISMATCH':
        return 'You can only offboard employees in your own tenant.';
      case 'USER_NOT_FOUND':
        return 'Employee not found.';
      default:
        return message;
    }
  }

  @override
  String toString() => 'AdminEmployeeException($code): $message';
}
